### PR TITLE
Add changes for dynamic environment support in choreo

### PR DESCRIPTION
**Pull request by VajiraPrabuddhaka**
### Purpose
Adding changes for supporting dataplane wise APIs fetching from control plane, using a dataplane ID provided in the configs.

- This will introduce the following additional configs under `controlPlane`

```toml
[controlPlane.dynamicEnvironments]
enabled = true
dataPlaneID = "$env{data_plane_id}"
```

- This would call the following endpoint of the control plane to fetch APIs at the startup, if the `enabled = true` in above configs

```
/retrieve-runtime-artifacts?organizationId="acme_uuid"&dataPlaneId="dp1-identifier"&type="Envoy"
```

- When a deploy event receives, it would compare the data-plane ID included in the event with the one in configured and deploy only of the API event has the relevant data-plane ID configured in gateway.

Sample Event :
```
 {"apiId":0,"uuid":"63e3319882b3046a5cae7312","name":"Test_mocky_api","version":"1.0.0","provider":"c2cc481c-b41a-4b00-ad65-4ec7031221dc","apiType":"HTTP","gatewayLabels":["Raveen-Env-dxjj"],"associatedApis":[],"context":"/abf64c3b-258d-4746-8279-d6c5f63ea558/dxjj/test/1.0.0","envToDataPlaneIdMap":{"Raveen-Env-dxjj":"2ACD5FFB-F012-ED11-A1F9-002248543709"},"eventId":"3228f34c-da98-4f41-aa3e-54bbc352dbb6","timeStamp":1675839946094,"type":"DEPLOY_API_IN_GATEWAY","tenantId":0,"tenantDomain":"abf64c3b-258d-4746-8279-d6c5f63ea558"}
```
### Issues

Fixes #

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/pkg/messaging/event_types.go [hunk 1]
```diff
@@ -104,22 +104,26 @@ type Event struct {
...
+	APIType             string                     `json:"apiType"`
 	Event
 	// TODO: (VirajSalaka) Remove this when the event is fixed from APIM side
 	Version string `json:"version"`
 	Context string `json:"context"`
 	Name    string `json:"name"`
 }
 
+// GatewayEnvToDataPlaneIDMap is used to represent gateway environment to data-plane ID mapping in APIEvent
+type GatewayEnvToDataPlaneIDMap map[string]string
```
Comment (VajiraPrabuddhaka) [inline adapter/pkg/messaging/event_types.go:125]: don't need this, we may use `map[string]string` directly inside APIEvent

File: adapter/internal/synchronizer/apis_fetcher.go [hunk 2]
```diff
@@ -160,34 +161,48 @@ func MergeDeployedRevisionList(deployedRevisionList []*notifier.DeployedAPIRevis
...
-	//finalEnvs contains the actual envrionments that the adapter should update
+	//finalEnvs contains the actual environments that the adapter should update
 	var finalEnvs []string
-	if len(configuredEnvs) > 0 {
-		// If the configuration file contains environment list, then check if then check if the
-		// affected environments are present in the provided configs. If so, add that environment
-		// to the finalEnvs slice
+
+	if conf.ControlPlane.DynamicEnvironments.Enabled {
 		for _, updatedEnv := range updatedEnvs {
```
Comment (VajiraPrabuddhaka) [inline adapter/internal/synchronizer/apis_fetcher.go:176]: instead iterating through updatedEnvs, we can iterate through `envToDpMap`

**Review by VajiraPrabuddhaka — COMMENTED** [review]
Review suggestions by @VirajSalaka

File: adapter/pkg/messaging/event_types.go (hunk 1 above)
Comment (VajiraPrabuddhaka) [reply adapter/pkg/messaging/event_types.go]: fixed in https://github.com/wso2/product-microgateway/pull/3212/commits/7d325deba6d2def0dcc0baadc01b2db643061ee2

File: adapter/internal/synchronizer/apis_fetcher.go (hunk 2 above)
Comment (VajiraPrabuddhaka) [reply adapter/internal/synchronizer/apis_fetcher.go]: fixed in https://github.com/wso2/product-microgateway/pull/3212/commits/7d325deba6d2def0dcc0baadc01b2db643061ee2

**Review by VirajSalaka — APPROVED** [review]

**Review by pubudu538 — APPROVED** [review]

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] DataplaneE1 cluster : dev-deployment-v2 : 20230213.11

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20230213.11

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20230213.11

**Comment by choreo-cicd** [comment]
[succeeded]  : prod-deployment-v2 : 20230215.1

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20230215.1

**Pull request merged**